### ISSUE TITLE
`self` is always considered to be truthy in some circonstances

### ISSUE THREAD

**Issue by qexat** [microsoft/pyright#4439]
**Describe the bug**
Pyright seems to consider `self` as always truthy even if `__bool__` might be falsy in some cases.

**To Reproduce**
Write a class that implements the magic method `__bool__` in a way that it might return `False` in some accessible cases (such as below)
```py
class C:
    def __init__(self, value: int = 0) -> None:
        self.value = value

    def __bool__(self) -> bool:
        return self.value >= 0

    def method(self) -> None:
        while not self:
            print("DECREMENTED!")
            self.value += 1
```

**Current behavior**
Within the `while` loop, `self` is typed as `Never`, but for example, running `C(-2).method()` prints `DECREMENTED!` two times.

**Expected behavior**
`self` should be typed as `self: Self@C` or equivalent

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/43090614/211867823-3a790bf9-c5f3-4865-a620-47d14b3ed8e3.png)
![image](https://user-images.githubusercontent.com/43090614/211867971-fe4455ac-83e6-4516-a436-2e84df68e170.png)

**VS Code extension or command-line**
Pyright is being run as a VS Code extension (Pylance, last release).

**Additional context**
n/a

**Comment by qexat** [comment]
Btw, I should add that removing the default value of `value` in `__init__` doesn't change the evaluation

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.289, which I just published. It will also be included in a future release of pylance.

**Comment by qexat** [comment]
Thank you for the quick fix!
